hw/bus/include/bus/bus_driver.h

```diff
@@ -58,7 +58,11 @@ struct bus_dev_ops {
                  uint16_t length, os_time_t timeout, uint16_t flags);
     /* Write data to node */
     int (* write)(struct bus_dev *dev, struct bus_node *node, const uint8_t *buf,
-                  uint16_t length, os_time_t timeout,  uint16_t flags);
+                  uint16_t length, os_time_t timeout, uint16_t flags);
+    /* Write and read data to node */
+    int (* write_read)(struct bus_dev *dev, struct bus_node *node, const uint8_t *wbuf,
```

Comment:
I think this should be a driver not a HAL.  I am not crazy about DMA being used within the HAL in general.

Reply:
I see what you mean but the current HAL has issues working in blocking mode, in an effort to resolve those issues, work was being done to get blocking mode working reliably with a solution for clock stretching which is almost impossible to resolve without using TWIM.

Reply:
extending HAL means we should also implement this for other platforms (not a big problem probably) but more generic problem here is that imo HAL should just abstract hardware in a simple way while this addition is not because of I2C in genreal, but because of specific chip.

I can see that in Nordic SDK they have indeed tx+rx and tx+tx implemented, but also I can see in spec that TWIM can use suspend/resume to stretch the clock and thus should work with separate write and read just like TWI does. But I did not try it so perhaps there are some problems which are not covered by spec. anyway, I'm not entirely against extending HAL but I'd use it as a last resort here.

Reply:
I think we can add a more advanced driver interface for this, with a naive implementation using the HAL, and a HW specific one targeting Nordic.  I definitely agree the API needs extending, but I don't think it needs extending for the HAL.

If we want, then sensor interface can rely on driver, and it will work reliably for cases when it's just HAL, but if efficiency is desired a native implementation can happen.  I assume most of what's done in this PR can be done naively?

Reply:
I see the point about extending the HAL for all mcus. not a very big deal per say, we can combine write + read for mcus that do not have such support or even by default just have write + read to begin with and over time as the mcus get used, support for a combined transaction will get added. Just thinking on the fly. I haven't really thought about other MCUs particularly.

Reply:
@sterlinghughes I do not understand. So, the way it is structured currently App->Sensor Layer->Sensor Drivers->Bus Driver->HAL (generic) -> HAL (MCU specific).

We could make decisions based on the fields in the sensor interface (sensor_itf) but the MCU specific HAL implementation would still remain broken as per the current implementation. How about adding MCU specific blocking/non-blocking approaches to the bus driver instead ? Only APIs will reside in the HAL, reset everything will reside in the bus drivers. 

Oh, and also in case if you are talking about fixing an issue about one sensor in the sensor layer with an MCU specific approach, that could be possible, but that means clock stretching needs to be handled in the sensor layer as opposed to the I2C driver/HAL.

Reply:
not sure if this is what @sterlinghughes had in mind, but with bus driver I can just implement `i2c_nordic_twim` driver and let I2C bus use it instead of `i2c` (which uses HAL) - we have untouched HAL, working TWIM on Nordic and drivers won't even notice something has changed.